### DISCUSSION TITLE
OnPopUpDestroyed caused the whole application shutdown

### DISCUSSION THREAD

**Discussion by dzunglean** [cefsharp/CefSharp#4534] [General]
Hello,
I am implementing a tab host control for the Popup window using LifeSpanHandler and facing the issue that the application is shut down every time the popup close (window.close()).

The root cause may come from DoClose() method since  onPopupDestroyed?.Invoke(control, browser) call is run in a separate thread hence the control.dispose() happen before the containing form could remove the ChromiumHostControl.

https://github.com/cefsharp/CefSharp/blob/0c7bc3f9b68e6196b2f2e60a2326aa8dadb7dfa2/CefSharp.WinForms/Handler/LifeSpanHandler.cs

**Comment by aimcosoftware** [comment]
There is an example here: https://github.com/cefsharp/CefSharp/wiki/Creating-user-tabs-or-popup-windows-in-Winforms

**Reply by dzunglean** [reply]
Thank. It is helpful.

**Comment by amaitland** [comment]
> I am implementing a tab host control for the Popup window using LifeSpanHandler and facing the issue that the application is shut down every time the popup close (window.close()).

What does your code look like (no **images of code**)? It's impossible to say what's actually happening without an example.

> The root cause may come from DoClose() method since onPopupDestroyed?.Invoke(control, browser) call is run in a separate thread hence the control.dispose() happen before the containing form could remove the ChromiumHostControl.

How did you come to that conclusion exactly? The call is invoked on the `UI thread` in a `sync` fashion to ensure you can remove the control from it's parent. The `Control.Dispose` call is made in the same action, so it's always going to happen after the `onPopupDestroyed` has invoked. Nothing looks out of place on first glance.

You'll need to provide a detailed example.

**Reply by aimcosoftware** [reply]
I am using very similar code to the Wiki, and I never get any unexpected behaviour.

I don't use `OnPopupDestroyed()`, I just cleanup inside `DoClose()`

**Reply by dzunglean** [reply]
Yes, it seems that if I call ChromiumHostControl.Dispose(), the whole application will be shutdown.

**Reply by dzunglean** [reply]
_tabControl.RemoveChromiumHost();
            _tabControl = null;
            MainPanel.Controls.Clear();
            OnApplicationSelected(_applications.First(x => x.ApplicationType == _currentApp));
            **popupControl.Dispose();**

**Reply by amaitland** [reply]
Code example is too minimal for me to know what's going on. There's no mention of `OnPopupDestroyed`.

Example usage at https://github.com/cefsharp/CefSharp/wiki/General-Usage#winforms---hosting-popup-using-tab-control
